sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py

```diff
@@ -4,61 +4,99 @@
 # Licensed under the MIT License.
 # ------------------------------------
 
-from base64 import b64encode
-import re
 from typing import TYPE_CHECKING
 
-from azure.core.pipeline.policies import SansIOHTTPPolicy
+from azure.core.exceptions import ServiceRequestError
+from azure.core.pipeline.policies import HTTPPolicy
+
+from ._exchange_client import ACRExchangeClient
 
 if TYPE_CHECKING:
-    from azure.core.pipeline import PipelineRequest
+    from azure.core.credentials import TokenCredential
+    from azure.core.pipeline import PipelineRequest, PipelineResponse
+    from typing import Optional
+
 
+def _enforce_https(request):
+    # type: (PipelineRequest) -> None
+    """Raise ServiceRequestError if the request URL is non-HTTPS and the sender did not specify enforce_https=False"""
 
-class ContainerRegistryUserCredential(object):
-    """Credential used to authenticate with Container Registry service"""
+    # move 'enforce_https' from options to context so it persists
+    # across retries but isn't passed to a transport implementation
+    option = request.context.options.pop("enforce_https", None)
 
-    def __init__(self, username, password):
-        self._user = username
-        self._password = password
+    # True is the default setting; we needn't preserve an explicit opt in to the default behavior
+    if option is False:
+        request.context["enforce_https"] = option
 
-    def get_token(self):
-        token_str = "{}:{}".format(self._user, self._password)
-        token_bytes = token_str.encode("ascii")
-        b64_bytes = b64encode(token_bytes)
-        return b64_bytes.decode("ascii")
+    enforce_https = request.context.get("enforce_https", True)
+    if enforce_https and not request.http_request.url.lower().startswith("https"):
+        raise ServiceRequestError(
+            "Bearer token authentication is not permitted for non-TLS protected (non-https) URLs."
+        )
 
 
-class ContainerRegistryUserCredentialPolicy(SansIOHTTPPolicy):
-    """HTTP pipeline policy to authenticate using ContainerRegistryUserCredential"""
+class ContainerRegistryChallengePolicy(HTTPPolicy):
+    """Authentication policy for ACR which accepts a challenge"""
 
-    def __init__(self, credential):
+    def __init__(self, credential, endpoint):
+        # type: (TokenCredential, str) -> None
+        super(ContainerRegistryChallengePolicy, self).__init__()
+        self._scopes = "https://management.core.windows.net/.default"
         self._credential = credential
+        self._token = None  # type: Optional[AccessToken]
+        self._exchange_client = ACRExchangeClient(endpoint, self._credential)
 
-    @staticmethod
-    def _update_headers(headers, token):
-        headers["Authorization"] = "Basic {}".format(token)
+    def _need_new_token(self):  # pylint: disable=no-self-use
+        # type: () -> bool
+        return True
```

Comment:
Is this awaiting an expanded implementation?

Reply:
Yes. The first preview release will not have any caching, but in the future this is where the caching implementation will go.